### PR TITLE
[4.0] com_finder: Filtering common & numeric words and language fix

### PR DESCRIPTION
**Pull request by Hackwar**
This PR filters common words and numeric terms from the index and the search query and also fixes a multilanguage issue in com_finder.

Searching for ```I, and, or, it,...``` in com_finder will give you lots of results, but might not be really usefull. So it is common for index based searches, to filter such common words both from the index (in order to keep it smaller) and from the search queries in order to only search for the relevant words. With #20781 I provided a way to add language specific sets of common words, this PR provides a switch to filter out these common words and a second one to filter out numeric terms.

I also found an issue with the multi-language capabilities of com_finder. Up till now, com_finder did not differentiate between terms from different languages. As long as they were written the same, they were okay. However with the stemming and the common words per language, we run into issues here. All of a sudden, a common word might still appear in the index for a content item, because it is correctly filtered out in english, but in the french article that has an english citation, it is still included. Thus it all of a sudden reappears and has some (albeit minor) negative effects on the search results. We are also having issues with different words from different languages which have the same stemmed form. Thus this change now does not silently drop identical words from different languages, but instead inserts each duplicate term with its specific language flag.

Please note that this PR has DB updates which need to be applied before testing.

### CODE REVIEW

File: administrator/language/en-GB/en-GB.com_finder.ini [hunk 1]
```diff
@@ -14,6 +14,8 @@ COM_FINDER_CONFIG_ENABLE_LOGGING_LABEL="Enable Logging"
 COM_FINDER_CONFIG_EXPAND_ADVANCED_LABEL="Expand Advanced Search"
 COM_FINDER_CONFIG_FIELD_OPENSEARCH_DESCRIPTON_LABEL="OpenSearch Description"
 COM_FINDER_CONFIG_FIELD_OPENSEARCH_NAME_LABEL="OpenSearch Name"
+COM_FINDER_CONFIG_FILTER_COMMONWORDS_LABEL="Filter common words from index?"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_finder.ini]: Labels should be short and they should not be questions

File: administrator/components/com_finder/helpers/indexer/indexer.php [hunk 2]
```diff
@@ -495,6 +496,15 @@ private function tokenizeToDbShort($input, $context, $lang, $format, $count)
 	 */
 	protected function addTokensToDb($tokens, $context = '')
 	{
+		static $filterCommon, $filterNumeric;
+
+		if (is_null($filterCommon))
+		{
+			$params = ComponentHelper::getParams('com_finder');
+			$filterCommon = (bool) $params->get('filter_commonwords', false);
```
Comment (Quy) [inline administrator/components/com_finder/helpers/indexer/indexer.php]: Why not keep it as `0` and `1` as in the XML without having to convert to boolean?

File: administrator/language/en-GB/en-GB.com_finder.ini (hunk 1 above)
Comment (Hackwar) [reply administrator/language/en-GB/en-GB.com_finder.ini]: Better?

File: administrator/components/com_finder/helpers/indexer/indexer.php (hunk 2 above)
Comment (Hackwar) [reply administrator/components/com_finder/helpers/indexer/indexer.php]: Why not convert it to boolean? We are checking this quite often and in best case we thus prevent some casting to boolean each time. Honestly there is no specific reason that I'm casting to bool here.

Comment (Quy) [reply administrator/components/com_finder/helpers/indexer/indexer.php]: For consistency, here one place where no casting `if ($params->get('filter_commonwords', 0) && $token->common)`

File: administrator/language/en-GB/en-GB.com_finder.ini [hunk 3]
```diff
@@ -14,6 +14,8 @@ COM_FINDER_CONFIG_ENABLE_LOGGING_LABEL="Enable Logging"
 COM_FINDER_CONFIG_EXPAND_ADVANCED_LABEL="Expand Advanced Search"
 COM_FINDER_CONFIG_FIELD_OPENSEARCH_DESCRIPTON_LABEL="OpenSearch Description"
 COM_FINDER_CONFIG_FIELD_OPENSEARCH_NAME_LABEL="OpenSearch Name"
+COM_FINDER_CONFIG_FILTER_COMMONWORDS_LABEL="Filter common words"
```
Comment (Quy) [inline administrator/language/en-GB/en-GB.com_finder.ini]: Labels should be in proper case.

**Comment by Hackwar** [comment]
And now?

**Comment by TobsBobs** [comment]
Can this be tested without #20781?

**Comment by Hackwar** [comment]
YesAm 11.08.2018 13:47 schrieb Tobi <notifications@github.com>:Can this be tested without #20781?

File: administrator/components/com_finder/helpers/indexer/indexer.php [hunk 4]
```diff
@@ -493,6 +494,15 @@ private function tokenizeToDbShort($input, $context, $lang, $format, $count)
 	 */
 	protected function addTokensToDb($tokens, $context = '')
 	{
+		static $filterCommon, $filterNumeric;
+
+		if (is_null($filterCommon))
+		{
+			$params = ComponentHelper::getParams('com_finder');
+			$filterCommon = $params->get('filter_commonwords', false);
```
Comment (Quy) [inline administrator/components/com_finder/helpers/indexer/indexer.php:502]: Change `false` to `0` to be consistent with XML. Repeat next line.

File: administrator/components/com_finder/helpers/indexer/query.php [hunk 5]
```diff
@@ -1269,7 +1338,8 @@ protected function getTokenData($token)
 		{
 			// Add the term to the query.
 			$query->where('(t.term = ' . $db->quote($token->term) . ' OR t.stem = ' . $db->quote($token->stem) . ')')
-				->where('t.phrase = 0');
+				->where('t.phrase = 0')
+				->where('t.language IN (\'*\',' . $query->q($token->language) . ')');
```
Comment (Quy) [inline administrator/components/com_finder/helpers/indexer/query.php:1341]: Change to `$db->quote($token->language)`?

**Comment by wilsonge** [comment]
Thankyou!

**Pull request merged**